---
title: "Workshop Mutation PSPG 245B"
author: "Lee and Martell"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output:
  html_document:
    df_print: paged
editor_options:
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# PSPG 245B
In this workshop you will learn about the fundamentals of mutation identification in cancer. By the end of the course you should be able to generate your own full R markdown for your assigned TCGA ID, including plots and identification of actionable mutations. Please see the the accompanying PDF for more information. 

## the following are required for this workshop
<div style="background-color:#b6d2db">

Software requirements  

* we assume that you are running > R 3.5 and have the below packges install. 
* use BiocManager::install() when possible. 
    + cgdsr
    + ggplot2
    + knitr
    + kableExtra
    + dplyr
    + VennDiagram

Please make sure to either clone the github repo or download the dropbox folder  

* https://www.dropbox.com/sh/lvaqmurdozncnxw/AADWRtK8woO4MjXIvGqcUcpXa?dl=1

* https://github.com/ahdee/workshop.mutation.2019

In this workshop we will be mainly using cbioportal's API to download the data we need. Thus it is important that you have access to the internet.

* We will be using the TCGA study on Breast Invasive Carcinoma (TCGA, Cell 2015) 
* Each student should had been assigned a TCGA ID prior to this.  

</div>


```{r, include=T, echo=T, message=FALSE, warning=FALSE, fig=TRUE}

# load require package and "source" an auxillary R script design for this workshop.  
# think of this as a cheat code. 

library ( cgdsr)
library ( ggplot2)
library(knitr)
library(kableExtra)
library ( dplyr)
library(VennDiagram)
source("auxi.R")



# initiate cbioportal 
cbiop = CGDS("http://www.cbioportal.org/")

```

# Lets learn how communicate with cbioportal and study what is available. 
* This portal is hierarchically structured. 
    + Head  List of all the cancer studies. 
    + identify the study you want
    + list all the "cases" for that study, a case is a description of both data available for the study and/or subsets of certain phenotype
	+ With each case you can get the list of all the associated TCGA identify
	+ which you can then use to get all the clincal data assocaited with the TCGA samples. 
	+ moreover you use the study to get specific genetic profile ( that is all the datasets available for downloading ). 
	

```{r }

# Get list of cancer studies at server
studies = getCancerStudies(cbiop)

# This will return a table with a list of available cancer studies to draw from 
k2( head ( studies, 10) , "example list of data available in cbioportal")

# note: k2 is one of those cheat codes to make tables print out nicely 
# ? what is the "10" for in the function above. 

# lets find out how many studies are available to the public
dim ( studies )

# ? how do you think we can use this to look for our dataset
# ? Because there are > 200 studies how do you think we can easily find what ware are looking for. 


# Find the dataset you want. 
# According to the syllabus we are looking for TCGA study on Breast Invasive Carcinoma (TCGA, Cell 2015) 
# However lets first search for all the breast studies available so that we can look to see which one most closely matches. 

breast = studies[ grepl("Breast", studies$name, ignore.case = T ), ]

# ? what is the T for what if you use lower case instead of upper case

k2(breast, "list of breast related studies")

### First BONUS question! try to find if there are other cancer data you might interested in, eg Lung Cancer

# from the table we now know that the study brca_tcga_pub2015 is what we need

brca.study = "brca_tcga_pub2015"

# note: each institution has different ways of organizing their data, here we first 
# identify the case studies and then get the cancer_study_id we need for a specific study 
# now that we have a case id we can search to see what types of samples are available AND types of data 
# this is your caselist - this are all the available data availabe for your specific study 

mycaselist = getCaseLists(cbiop,brca.study)
# first lets study what is in the case list.  
k2 ( head ( mycaselist[ ,c("case_list_id",	 "case_list_name",	 "case_list_description",	 "cancer_study_id"	)], 50), "list of cases for the BRCA CELL 2015 study" ) 

# ? can you tell us what each of the column mean based on what is outputed? 
# ? Which data do you think is most relevant this workshop? 

# here we see that brca_tcga_pub2015_all is probably the best to use because it includes ALL Complete Tumors 
# compared to brca_tcga_pub2015_her2pos which are only samples that are Her2-positive breast tumors 

case.list.id = "brca_tcga_pub2015_all"


# ok now using All Complete Tumors lets see if your sample is present. 

mysample = mycaselist[ mycaselist$case_list_id == case.list.id, ]$case_ids
mysample = unlist ( strsplit(mysample, " ") )

# ? mysample variable now contains all the TCGA samples in this study, can you check if your sample is exists? 

# here is the list of samples. Do you see your sample in here? 

cat ( samples, sep="\n")

# here I chose a random sample to test 
# IMPORTANT! here you need to use your own assigned TCGA ID 

myid = "TCGA-C8-A131-01"
mysample[mysample== myid]

# here is another way
intersect ( mysample, myid)
# yet another 
mysample[grepl(myid, mysample)]

#? can you think of any reason to use grep over another? 

### Now lets check if it matches with what is reported: 817 
length ( mysample)

# lets get clinical some clinical data 
myclinicaldata = getClinicalData(cbiop,getCaseLists(cbiop,brca.study)[1,1])
# clean up the names a bit 
myclinicaldata$pid = gsub ( "\\.", "-", row.names(myclinicaldata))
# lets get the clinical information for your TCGA id. 
k2 ( t ( myclinicaldata[myclinicaldata$pid %in% myid, ] ))

# important fields to consider. 
# AGE, AJCC_PATHOLOGIC_TUMOR_STAGE, Ethnicity
# DFS ( disease free), OS ( overall survival)



### Bonus Lets have some fun here, even with just a table of clinical data we can do some investigation. 
# For example how would you figure out if one of your sample Triple-negative or HER2+. 
# hint look at the caselist and work backwards using the intersect function

k2 ( head ( mycaselist[ ,c("case_list_id",	 "case_list_name",	 "case_list_description",	 "cancer_study_id"	)], 50), "list of cases for the BRCA CELL 2015 study" ) 

t3.study = "brca_tcga_pub2015_trineg"
t3 = mycaselist[ mycaselist$case_list_id == t3.study, ]$case_ids
t3 = unlist ( strsplit(t3, " ") )


her2p.study = "brca_tcga_pub2015_her2pos"
her2p = mycaselist[ mycaselist$case_list_id == her2p.study, ]$case_ids
her2p = unlist ( strsplit(her2p, " ") )

er_neg.study = "brca_tcga_pub2015_erneg"
er_neg = mycaselist[ mycaselist$case_list_id == er_neg.study, ]$case_ids
er_neg = unlist ( strsplit(er_neg, " ") )


er_POS.study = "brca_tcga_pub2015_erpos"
er_POS = mycaselist[ mycaselist$case_list_id == er_POS.study, ]$case_ids
er_POS = unlist ( strsplit(er_POS, " ") )


# answer. 
t3[t3==myid]
her2p[her2p==myid]

# Also suppose you ask how many are HER2 positive but ER +? 

venn1 = data.frame ( her2p= length ( her2p)
                    , er_neg= length ( er_POS )
                    , int= length ( intersect(er_POS,her2p))  
                        )
names ( venn1 )[1:2] = c("Her2+","ER+") 

venn.this ( 
    venn1, 
    type = 2, cp= c("#3C8A9B" , "#9B445D"), 
    dgg=180
) 

# 

age.both = intersect (er_POS,her2p )
age.both=myclinicaldata[ myclinicaldata$pid %in% age.both, ]$AGE
age.Her2only.noER = setdiff( her2p, er_POS)
age.Her2only.noER = myclinicaldata[ myclinicaldata$pid %in% age.Her2only.noER, ]$AGE

temp = rbind ( 
data.frame ( age = age.both, type= rep("both", length(age.both)) )    
,data.frame ( age = age.Her2only.noER, type= rep("Heronly", length(age.Her2only.noER)))   
)

median ( age.Her2only.noER )
median ( age.both )

ggplot( temp, aes(x=age, fill=type)) +
    geom_histogram( color="#e9ecef", alpha=0.6, position = 'identity') +
    scale_fill_manual(values=c("#69b3a2", "#404080")) 

#? BONUS question to take home. As a pharmacologist how can you use DFS and OS? For example 
# the histogram above only shows age of onset.  However, question what if you hypothesize
# that having HER2 positive is much worse of survival then having both HER2 and ER+? 



# Now lets see what information is available for your study 
mygeneticprofile = getGeneticProfiles(cbiop,brca.study)
k2 ( mygeneticprofile, "types of data available for chosen case" )


# so looking at that we now know it contains several interesting modalities.  Lets try the mutation
mutation = mygeneticprofile[grepl( "Mutation", mygeneticprofile$genetic_profile_name), ]$genetic_profile_id 

# ? what other modalities do you think may be of interest
# anyone want to guess what "probe most anti-correlated with expression means"
cna = mygeneticprofile[grepl( "HM450", mygeneticprofile$genetic_profile_name),1 ]

exp = mygeneticprofile[grepl( "V2 RSEM", mygeneticprofile$genetic_profile_name),1 ]

```



# Now that we have a basic understanding of what is available lets go and download the actual data. 

```{r}

# lets collect this through a loop so not to overwhelm the system
# its important to be mindful of resources when mining data.  We need to play nice and consider the server load. 
# ? what do you think we can do to miminize server load? 
# cancer.gene was previosly imported so that we are only going to look for gens asscoated with cancer. This is from various sources 
length ( cancer.gene)

total = ceiling ( length( cancer.gene)/100  )  *100
mutations = data.frame (   stringsAsFactors = F )
expression = data.frame (   stringsAsFactors = F )
e = 1
for(i in seq(from=200, to=total, by=200)){
   
    if ( i > length(cancer.gene)){
        i = length(cancer.gene)
    }
    
    print ( paste ( e, i ))
    temp = getMutationData(cbiop , brca.study, mutation, cancer.gene[e:i])

    
    mutations = rbind ( mutations, temp)
    
        e2 = 1
        exp.temp2 = 1
        for(i2 in seq(from=200, to=length ( myclinicaldata$pid), by=200)){
        exp.temp = getProfileData(cbiop , cancer.gene[e:i]
                              , exp
                              , getCaseLists(cbiop,brca.study)[1,1] , myclinicaldata$pid[e2:i2])
        if ( exp.temp2 == 1){
            exp.temp2 =   t ( exp.temp)
        }else {
            exp.temp2 =  cbind ( exp.temp2,  t ( exp.temp)) 
        }
        e2 = i2+ 1 
        Sys.sleep (1) # lets give the system a break
        }
        
    e = i + 1    
    expression = rbind ( expression, exp.temp2)    
    
    Sys.sleep (2) # lets give the system a break
    
}
# create coordinates
mutations$igv = paste ( mutations$chr, mutations$start_position, mutations$end_position, mutations$reference_allele, mutations$variant_allele)

# there is an error in the mutation table.  The links are obsolete.  Here we fix it. 
# by replacing the older url with this: http://mutationassessor.org/r3

mutations$xvar_link = gsub('getma.org','http://mutationassessor.org/r3', mutations$xvar_link)
mutations$xvar_link_msa = gsub('getma.org','http://mutationassessor.org/r3', mutations$xvar_link_msa )
mutations$xvar_link_pdb = gsub('getma.org','http://mutationassessor.org/r3', mutations$xvar_link_pdb  )

expressionc= expression[complete.cases(expression), ]
samplesnice = gsub ( "-",".", samples)
expressionc[ row.names (expressionc ) %in% c(  "TP53" , "KRAS"), samplesnice ]

```

# lets study the mutation table

```{r}

dim ( mutations )

# lets take a few minutes here to go over the different fields. 
names ( mutations )
?getMutationData

unique( mutations$mutation_type)
# ? based on the lecture previously can identify the different mutation_type
# ? based on the lecture previously can you figure out what each of these mean?  
# for example which of these is a nonsynonymous substitution
# ? for druggability which one of these do you think its more likely to be useful

# ? can you guess what is functional_impact_score
k2 ( data.frame ( table ( mutations$functional_impact_score)  ), "impact scores " )

# Bonus make sure that your sample has mutations. 

nrow ( mutations [ mutations$case_id ==  myid , ] )

# ? if you have more than one sample, say you want to check all 10 samples in the class, how would you do this? 
# ? this is important, for example previously we were able to subset for Triple-negative breast tumors

samples

final =  mutations [ mutations$case_id %in% samples, ]
unique ( final$case_id)

# so as you can see above there are a total of 10 unique samples. 

# lets clear up junk 
gc()
                          

```


# Now lets talk a bit about how to identify pathogenic mutations. 
## Can you recall some of the things that were mentioned previously? 

```{r}

# One easy approach is to look for "cancer" related genes that had already been implicated in cancer.
# ? how do you think we can do that? 

k2 ( cancer.list[17:25, ], "cancer list")

# ? can you recall what is exome sequencing and why its important?

# lets check this list to make sure genes that are relevant to breast cancer exists 

imp = as.character ( cancer.list[grepl("^BRCA|^ATM$|^BARD1$|^CDH1$|^CHEK2$|^NBN$|^NF1$|^PALB2$|^PTEN$", cancer.list$gene), ]$gene )

# ? why did we add ^ in front and $ for only some genes and not others?

k2 ( cancer.list[cancer.list$gene %in% imp, ], "breast cancer genes")

# ? based on this which panel do you think is weakest for breast cancer and why do you think that is the case? 

# Bonus: can you check if BRCA is present in any of the mutations we just downloaded? 

k2 ( head ( mutations[ grepl("^BRCA", mutations$gene_symbol), ] ), "BRCA gene mutations" )


# lets see how many genes do we have all together 
nrow ( cancer.list )

### BONUS can you check if your favorite gene is in here? 

# Another way we can detect important mutations is to see if other resources are available. # Here we use cosmic a manaully curated list of recurrent mutations in cancer. 

k2 ( head ( cosmic.70[ grepl("breast",  cosmic.70$description ), ] ), "breast cancer in cosmic")
# ? what do you think the 70 means?


# lets integrate cosmic into our mutation table

mutations = merge ( mutations, cosmic.70, by="igv", all.x=T, all.y=F )

# ? why do you think all.x = T and all.y=F 

# let us find out how many mutations we have that is also found in cosmic 

nrow ( mutations[ ! is.na ( mutations$description ),  ] ) / nrow (mutations)

# ? wow thats a huge ratio , why do think the ratio is so high? 

# BONUS check if cosmic contain BRCA mutations.  This is a good sanity check. 

k2 ( head ( mutations[ grepl("^BRCA", mutations$gene_symbol), c("amino_acid_change","chr.x","start_position","end_position","reference_allele", "variant_allele",
"gene_symbol","case_id","mutation_status","mutation_type","description")] ) )
# BONUS go to https://cancer.sanger.ac.uk/cosmic and compare results of line 4 
# https://cancer.sanger.ac.uk/cosmic/search?q=3813722
# ? does everything match
# ? why do you think the the coordinates are off by 1


# BONUS can you find all BRCA mutation that is Missense_Mutation AND found is cosmic in the mutation db?


# hint, mutations$mutation_type == "Missense_Mutation"
```

# Now that the data has been succesfully donwloaded lets study your breast cancer set. 

```{r}



# lets tabulate the types of mutations mostly seen with Breast cancer 

mutation.type = data.frame ( table ( mutations$mutation_type))

mutation.type = mutation.type[ order ( mutation.type$Freq), ]

mutation.type$Var1 = factor ( mutation.type$Var1, levels = mutation.type$Var1)

ggplot(mutation.type, aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "BRCA mutation types")





# ? seems like missense mutation is the most recurrent.  Can you think of any way to clean up the data 

# ? in the lecture we talked about depth.  So lets calculate this. 

mutations$dp = mutations$variant_read_count_tumor + mutations$reference_read_count_tumor 

 ggplot(mutations, aes(x=dp)) + 
  geom_density( color="darkblue", fill="steelblue" ) + geom_vline(xintercept = 20, linetype="dotted", 
                color = "grey", size=1.5) + ggtitle ( " density plot, total depth")

# ? by looking at this plot we can see that there is a good amount of mutations that are under 20 
# ? how would you remove this?

# before  
dim ( mutations )
# after 
mutations = mutations[ mutations$dp > 20, ]
dim (mutations )


# lets calculate allele freqeuncy here as well 
mutations$af = mutations$variant_read_count_tumor / mutations$dp

 ggplot(mutations, aes(x=af)) + 
  geom_density( color="darkblue", fill="#e8975a" ) + geom_vline(xintercept = .1, linetype="dotted", 
                color = "grey", size=1.5) + ggtitle ( " density plot, allele freqeuncy")

# ? what do you think it means when af is higher for a particular mutation 
# Bonus pick the top 10 mutations ranked by af 

mutations = mutations[ order ( - mutations$af ) , ]

k2 ( head ( mutations, 10) , "top 10 variant ranked by AF")

# lets further filter the mutations with any af > .1
mutations = mutations[ mutations$af > .1, ]
dim (mutations)


# plot 20 highest recurrent gene mutations

mutation.gene = data.frame ( table ( mutations$gene_symbol))

mutation.gene = mutation.gene[ order ( - mutation.gene$Freq), ]
mutation.gene$Var1 = factor ( mutation.gene$Var1, levels = rev ( mutation.gene$Var1) )

ggplot(mutation.gene [ 1:20, ] , aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "Most recurrent genes")


# as mentioned previously one reliable way to look for relevant mutations is to match it against what has been previosly observed and/or study. 

cosmic.mutation = mutations [ !is.na(mutations$description), ]

mutation.type = data.frame ( table ( cosmic.mutation$mutation_type))

mutation.type = mutation.type[ order ( mutation.type$Freq), ]
mutation.type$Var1 = factor ( mutation.type$Var1, levels = mutation.type$Var1)

ggplot(mutation.type, aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "BRCA mutation types found in cosmic")


# ? compare the two plots, did anything change?

# based on literature these are some of the genes involved in breast cancer. 
imp 

# lets study them and see what types of mutations corresponds most with these genes. 

mutation.brca = data.frame ( table ( mutations[ mutations$gene_symbol %in% imp, ] $mutation_type))

mutation.brca = mutation.brca[ order ( mutation.brca$Freq), ]
mutation.brca$Var1 = factor ( mutation.brca$Var1, levels = mutation.brca$Var1)

ggplot(mutation.brca, aes(Var1,Freq), label=Freq ) +
    geom_bar(aes(fill = Freq), stat="identity", position = "dodge") +
    coord_flip() +
    scale_fill_distiller(palette = "RdBu") + xlab("") + ylab("") +
    theme(strip.text.y = element_text(angle = 0), legend.position="none") +
    geom_text(aes(label=Freq), position=position_dodge(width=0.9), vjust=.4, hjust = .5, size=5) +
    theme(panel.grid.major = element_blank(), 
          panel.grid.minor = element_blank(),
          panel.background = element_blank(), 
          axis.line = element_line(colour = "black"), 
          text = element_text(size=16),  # size of label
          axis.text.x = element_text(angle=0, hjust=1) ) + ggtitle ( "Mutation-types in genes associated with breast cancer")

# lets assume that we don't know these genes 
# ? can you think of a way to find them based on what we have? 

# there are many like how we sorted by most frequent above. 
# another is to subset by what has already been observe in cosmic 

imp.brac = data.frame ( table ( mutations[ grepl( "breast", mutations$description), ]$gene_symbol ) )
imp.brac$fraction = imp.brac$Freq / nrow ( mutations)
imp.brac = imp.brac [ order ( -imp.brac$fraction) , ]

k2 ( head ( imp.brac, 10), "top cosmic genes found in breast cancer")

# ? how does this compare to our plot above 

# both analysis seem to show a lot of overlap eg that TP53 and PIK3CA are important.  Lets breakdown the mutation type a bit more for the top 5

top5 = as.character ( imp.brac$Var1[1:5] )
top5 = mutations[ mutations$gene_symbol %in% top5, ]$amino_acid_change
top5 = data.frame ( table ( top5 ) )
top5 = top5 [ order ( - top5$Freq), ]

# quick hack to add url, will not work with *

top5$url = paste0(url,top5$top5)
k2 ( head ( top5, 20), "top 20 recurrent mutations")



# part of studying mutations is look for clinical significance.  One way to do this is to match your mutations to actionable 
# targets


mutations$drug = paste ( mutations$gene_symbol, mutations$amino_acid_change)

action = drug[ drug$match %in% unique ( mutations$drug  ), ]
action.f = c("gene" ,"variant" , "disease", "drugs", "evidence_type", 
             "clinical_significance", "evidence_statement", "evidence_civic_url", "gene_civic_url" , "match"
)
k2 ( head ( action[ ,action.f], 20) ,"example of actionable mutations in the BRCA set" )

# Bonus can you figure out which samples have druggable targets?  
#? look at evidence type. Can yo figure out which of these are diagnostic vs targetable? 

k2 ( head ( action[action$clinical_significance == "Sensitivity/Response" ,action.f], 20) )



```

# HW at this point please use your own TCGA ID, study it using what you learn above.  Assess QC on your sample, study what genes are most common and which if any are targetable. 

```{r}
# here is example where I chose, myid = "TCGA-C8-A131-01"

 
actionid = drug[ drug$match %in% unique ( mutations[mutations$case_id == myid, ]  $drug  ), ]

actionid = actionid %>% dplyr::group_by(variant) %>%
        dplyr::summarise(
            gene = paste(unique ( gene ), collapse = "," ) ,
            disease = paste(unique ( disease ), collapse = "," ),
            clinical_significance =  paste(unique ( clinical_significance ), collapse = "," ),
            drugs = paste(unique ( drugs ), collapse = "," ) 
            
        ) %>%
        data.frame()

k2( actionid, myid)
 
```

